fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,123 @@
+    * Then 😉, once the response JSON is received and parsed,
+    *   - create a shallow copy of the todos array from state
+    *   - add the todo received from the server to the copied array
+    *   - replace the old todos array in state with the new one
+    *
+    * DO NOT MUTATE the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+    fetch('api/todos', {
+      method: 'POST',
+      headers: { 'content-type': 'application/json' },
+      body: JSON.stringify(newTodo)
+    })
+      .then(res => res.json())
+      .then(data => {
+        const todosArray = this.state.todos.slice(0);
+        this.setState({
+          todos: todosArray.concat(data)
+        });
+      })
+      .catch(err => console.error('Fetch failed!', err));
+  }
+
+  toggleCompleted(todoId) {
+    /**
+     * Find the index of the todo with the matching todoId in the state array.
+     * Get its "isCompleted" status.
+     * Make a new object containing ONE PROPERTY: the opposite "isCompleted" status.
+     * Use fetch to send a PATCH request to `/api/todos/${todoId}`
+     * Then 😉, once the response JSON is received and parsed,
+     *   - create a shallow copy of the todos array from state
+     *   - replace the old todo with the todo received from the server
+     *   - replace the old todos in the state with the new one (you know the index).
+     *
+     * NOTE: "toggle" means to flip back and forth, so clicking a todo
+     * in the list repeatedly should "toggle" its isCompleted status back and forth.
+     *
+     * DO NOT try to calculate the index of the todo by subtracting 1 from the id.
+     *
+     * DO NOT MUTATE the original state array, nor any objects within it.
+     * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+     *
+     * TIP: Be sure to SERIALIZE the updates in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+    const todos = this.state.todos;
+    for (let i = 0; i < todos.length; i++) {
+      if (todos[i].todoId === todoId) {
+        const currentIndex = i;
+        const toggleUpdate = { isCompleted: !todos[currentIndex].isCompleted };
+        fetch(`/api/todos/${todoId}`, {
+          method: 'PATCH',
+          headers: { 'content-type': 'application/json' },
+          body: JSON.stringify(toggleUpdate)
+        })
+          .then(res => res.json())
+          .then(data => {
+            const todosArray = this.state.todos.slice(0);
+            todosArray[currentIndex].isCompleted = toggleUpdate.isCompleted;
```

Comment:
This is actually modifying the object at `currentIndex`. Instead, just swap it out with `data`.